In [1]:
import os
import gsw
import numpy as np
%matplotlib inline

# Paleo modeling


# Model setup

Simulations are run using the COAWST implementation of ROMS (version VERSION). The horizontal grid has been used in several studies (Hetland and Campbell, 2007; Zhang et al, 2014) and output is used operationally by the Texas General Land Office and NOAA's Office of Response and Restoration.

## Boundaries
All boundaries are closed. Walls are no-slip and bottom friction is used with (WHAT VALUES). There is no meteorological forcing.

## Bathymetry

The sea level was about 120 meters lower 19,000 years ago at the end of the last glacial period and the Mississippi river entered the Gulf at the Mississippi canyon.

![](images/bathy.png)

### Shifting
In the script `setup/bathy/bathy_shift.py`, I subtract 120 meters from all of the depths, mask out anything that went negative since it is now land, make everything have a minimum depth of 25 meters, fix a couple of weird holes in the mask by hand (not sure these ended up being important to model stability), and then update the masks for all grids. The output of this is `setup/bathy/grid-shifted.nc`.

### Smoothing
The script `setup/bathy/bathy_smooth.py` takes in `grid-shifted.nc` and smoothes the bathymetry to a Beckman/Haidvogel number of 0.4. It's hard to get the code installed, but I had it working on my old laptop and found that the different algorithms for smoothing tend to give fairly similar results, so I am just using "Positive" now.

### Vertical grid
The number of vertical levels is 20 with Vtransform=2 and Vstretching=4 to achieve a Haney (CITE) metric of 7.8.

## T and S initialization

### Data
The simulations are initialized with a uniform density gradient (with depth) calculated using the linear equation of state with temperature and salinity averaged over the profiles available in the World Ocean Atlas (1955-2012) from the deep Gulf of Mexico (Station number 28569 (B): 93.5W, 24.5N). The ODV [WOA 2013 data](https://odv.awi.de/en/data/ocean/world_ocean_atlas_2013/
) ([data link](https://odv.awi.de/fileadmin/user_upload/odv/data/WOA13/WOA13_1.00deg_1955-2012_Annual.zip
)) is saved into file `setup/initialization/data_from_WOA13_1.00deg_1955-2012_Annual_centerGOM.nc` and is shown here:

<img src="images/Profiles_AnalyzedValues_centerGOM.gif" width="70%">

### Present day background values
The background salinity and temperature are given by values from the bottom of the WOA profile where the values are more consistent: $S_0=35.0$ g kg$^{-1}$, and $T_0=4.2$ $^\circ$C. These are calculated from data arrays in `setup/initialization/initialization.py`, taking the bottom 20 bins (1950 to 3800 depth). I then calculate the background density using the equation of state and assuming pressure from the top of the water column:

In [2]:
# present day
S0 = 35; T0 = 4.2
rho0 = gsw.rho(S0, T0, 0)  # kg/m^3
print(rho0)

1027.63433399


### Paleo values
Based on Niall's work, the paleo temperature profile is then uniformly 2 degrees cooler and 3.5% more saline across whole water column. The background numbers are then $S_0=36.225$ g kg$^{-1}$, and $T_0=2.2$ $^\circ$C. I can similarly calculate the background paleo density:

In [3]:
# paleo
S0 = 36.225; T0 = 2.2
rho0 = gsw.rho(S0, T0, 0)  # kg/m^3
print(rho0)

1028.79903353


### Linear equation of state
I used the linear equation of state to calculate the paleo density profile, using the paleo background numbers. The linear equation of state is given by the equation below where $S_{COEF}=7.6\times 10^{-4}$ (g kg$^{-1}$)$^{-1}$ and $T_{COEF}=1.7\times 10^{-4}$ $^\circ$C$^{-1}$ from the ROMS model. 
$$ \rho = \rho_0 *\left[1.0 + S_{COEF} (S-S_0) - T_{COEF} (T-T_0)\right] $$

The resultant salinity and temperature profiles are on the left in the figure below.

In [4]:
Scoef = 7.6e-4; Tcoef = 1.7e-4

os.chdir('../setup/initialization')
%run make_initialization.py 1 0
os.chdir('../../notebook')

//anaconda/lib/python3.5/site-packages/matplotlib-2.0.2+4065.gde2755c-py3.5-macosx-10.5-x86_64.egg/matplotlib/mathtext.py:846: MathTextWarning: Substituting with a symbol from Computer Modern.
  MathTextWarning)


![profiles](../setup/initialization/WOA_profiles.png)

In case the figures aren't showing above, here is the image:

![](images/WOA_profiles.png)

### Density setup for simulation

After finding the density profile for the paleo initialization, I use the linear equation of state to calculate the equivalent temperature profile that would make that density profile if the salinity is uniformly equal to the background salinity. That is, $S=S_0$, so $ \rho = \rho_0 *\left[1.0 - T_{COEF} (T-T_0)\right] $ and 
$$ T = -\frac{\rho}{\rho_0 T_{COEF}} - T_0 $$

This is a numerical convenience so that salinity can be used as a tracer representing the input river water. The paleo density profile and temperature profile that gives the same density profile with no salinity changes are shown on the right above.

### Initialization file
The initialization file `ocean_ini.nc` is created by `python make_initialization.py 0 1`.

## River input

### Locations and structure

I have the river coming in around the Mississippi canyon (see bathymetry above). The river location options are visualized in the figures below, as made in `setup/river/river_locations.py`. Since the locations are defined on the u and v grid by the cell wall that the river water is actually exiting through, the proper grid labels are shown for each of the 3 main numerical grids.

<img src="images/masku.png" alt="ts" width="50%">
<img src="images/maskv.png" alt="ts" width="50%">

The river forcing file, `tglo_river_frc_2010-01-01.nc`, is made in `setup/river/river_frc_tglo.py`. This file is referenced in time to 'days since 1970-01-01 00:00:00' and starts on 2010-01-01, though the date isn't relevant for this simulation except to keep track of time passing. The time length is controlled in this file (currently a year of 30 day months) and the output is every 3 hours. The river is input over the full 20 vertical layers. The amount of river water in each layer is weighted by the thickness of the layer so that the profile is uniform over the water column. There are 24 input locations that the river water is divided across at each time step, and the amount out each cell face is weight by the surface area of the cell wall.

### Amount
The total volumne of water being input through the river is $1e14$ m$^3$, which is from Clarke et al 2004’s flood volume of 151,000 km^3 over 0.5 year. I used a Gaussian to represent the flood pulse:

$$Q(t) = A e^\left( - \frac{t-mid^2}{2w^2}\right)$$

where  $Q(t)$ is the flux in m^3/s, $A$ is the amplitude [m$^3$/s], $mid$ is the middle point of the gaussian, and $w$ is the width (0.15 * the 12 month runtime). I solve for A:

$$ \Sigma_t Q(t) = A * \Sigma_t e^{()} $$

and $\Sigma_t Q(t)$ is the volume, so $A = V / \Sigma_t e^{()} * dt  = 8561003 m^3/s$ (calculated in `setup/river/river_frc_tglo.py`).  Then I can solve for Gaussian.

<img src="images/river_pulse.png" alt="ts" width="20%">

### CFL and time step

The velocity for this flow rate is:

$$ u = Q_{max} / (\Delta x * \Delta z) = 460977 m^3/s / (5000 m * 25 m) = 3.7 m/s $$

The CFL condition is satisfied for

$$ t < 0.5 ∆x/u = 0.5 * 5000m / 3.7 m/s = 675 s $$

though the actual time step being used ends up being smaller.

### Density

The density of the river water is controlled by the salinity and the temperature is set to the background temperature. I choose the desired density of the river water and use the linear equation of state to solve for the necessary salinity.

## Key input parameters

* NTIMES == 86400  # 1 year simulation time
* DT == 360.0d0  # 6 min time step
* NHIS == 240  # daily output
* NDEFHIS == 86400  # all output in 1 year-long file  


# Initial results

I ran a series of four simulations with a range of river densities to span from the top to bottom of the water column. 

## low density, top of the water column

`new1`, river salinity = 33.9 (density 1026) (background salinity 36.225)
<img src="images/new1-180.png" alt="ts" width="30%">
<img src="images/new1-360.png" alt="ts" width="30%">

## medium-low density

`new2`, river salinity = 35.2 (density 1027)
<img src="images/new2-180.png" alt="ts" width="30%">
<img src="images/new2-360.png" alt="ts" width="30%">

## medium-high density

`new3`, river salinity = 36.5 (density 1028)
<img src="images/new3-180.png" alt="ts" width="30%">
<img src="images/new3-360.png" alt="ts" width="30%">

## high density, bottom of the water column

`new4`, river salinity = 37.8 (density 1029)
<img src="images/new4-180.png" alt="ts" width="30%">
<img src="images/new4-360.png" alt="ts" width="30%">


# Possible next directions

Several sources show a sea level rise 19,000 years ago. Clark et al (2004) the rise was at least 10 meters. Yokoyama et al (2000) say a rapid rise of 10-15 meters. Clark et al (20xx) the rise was at least 10 meters. Somebody was 5 meters (Mortlock?). So we are planning to use between a 5 and 10 meter sea level rise.

Given a surface area of the ocean of $3.61e8$ (from the way Mortlock calculated this for his work?), this gives a volume of water range of 1.805 to 3.61 million km$^3$ of seawater input into the ocean. (This is compared with the volume of river water I have previously run with: 100,000 km^3).

Niall has evidence in his cores that these inputs occurred over three different events, with time in between.

The time range is unknown because the dating methods are not precise enough. However, we do know that the range must be less than 500 years because of WHO'S data in WHAT PAPER. At least paper (WHICH?) also indicated a flood of size 10,000 km$^3$/year (IS THAT CORRECT?).

For reference, the typical present day Mississippi discharge per year is (from Wikipedia's mean number):

$$ 16792 m^3/s *(365 days/year) * 24 hours/day * 3600 seconds/hour) / 1000(m/km)^3 \approx 500 km^3/year$$

We probably only need to run one of these events. We could run simulations for several time lengths, with different approaches:

1. For a short, strong flood, I would need to use a very small time step but the simulation would be shorter.
1. For a long, weak flood, I could use a bigger time step and run a few targeted years during the long simulation time period since the time scale of the flood would be so long relative to the dynamics of the water being input into the Gulf.
1. Separately: keep in mind that we only need to run the simulation through the settling velocity time since that is the process we are comparing with the sediment data.

# Realistic volume simulations

We will run a spread of simulations: 1. 50% of a 5 meter sea level rise (this assumes there is input to the 5 meter sea level rise from other sources and represents a lower bound) to 2. 100% of a 10 meter sea level rise (this assumes all water is from this event and represents an upper bound). For both cases, I will divide the total volume by 3 to represent that there were 3 events in the cores.

1. Total volume: 1.805 million km$^3$ water * (1000 m/km)$^3$ = 1.805$\times$10$^{15}$m$^3$

   Take 50%: 0.5 * 1.805$\times$10$^{15}$m$^3$ = 9.025$\times$ 10$^{14}$ m$^3$
    
   Divide by 3 for number of pulses to get total volume per pulse in this scenario: 
   9.025$\times$ 10$^{14}$ m$^3$ / 3 $\approx$ 3$\times$10$^{14}$ m$^3$

1. Total volume: 3.61 million km$^3$ water * (1000 m/km)$^3$ = 3.61$\times$10$^{15}$m$^3$

   Take 100%
    
   Divide by 3 for number of pulses to get total volume per pulse in this scenario: 
   3.61$\times$ 10$^{15}$ m$^3$ / 3 $\approx$ 1.2$\times$10$^{15}$ m$^3$
   
For each simulation, these total volumes go into `paleo/setup/river/river_frc_tglo.py` as the `transport` variable.


## Temperature and salinity of river water

Also need to update the temperature and salinity to reflect the new background densities (which have already been changed above). As before, the density of the river water is controlled by the salinity. The desired density of the river water is 1028 kg m$^3$. For this river density, we can find the depth at neutral density in the water column and the corresponding temperature to set, along with using the linear equation of state to solve for the necessary salinity we get:

$$ S = \frac{1}{S_{COEF}} \left[\frac{\rho}{\rho_0} - 1 + T_{COEF} (T - T_0) \right] + S_0 $$

In [8]:
os.chdir('../setup/initialization')
%run make_initialization 0 0 --calc_river 1028
os.chdir('../../notebook')

neutral depth of river water: 460.70075757575756
surface temperature, to use for river water: 12 for depth 0.0
river salinity, from linear eqn of state: 37.3951759283


These values go into `setup/river/river_frc_tglo.py` under `river_temp` and `river_salt`.